### ISSUE TITLE
bitcoin/contrib/linearize/linearize-data.py. Premature end of data.

### ISSUE THREAD

**Issue by educob** [bitcoin/bitcoin#14986]
Hi. 

I have the mainnet blockchain updated (block 554227).
My app prcesses the whole blockchain and for that I need to read the blocks directly from disk as asking the node for them takes months (to get them and proccess them).

I run `python linearize-hashes.py linearize.cfg > hashlist.txt` with `max_height=554000` and then I run: `python linearize-data.py linearize.cfg`

It only processes block 0 cause in blk00000.dat the 2nd block starts with 0000000000000000 instead of f9beb4d9xxxxxxxx. After reading all blkxxxxxx.dat files linearize-data.py ends without finding block 1.

This same thing happened days ago in block around 120000 so I deleted the blockchain and started all over again (wasting almost 3 days).

Why does blk00000.dat have a 0000000000000000 after the 1st block?
How can I know where block 1 is?

Needless to say the node is capable of returning block 1 when calling client.getBlockByHash.

Please help.

ps: I think that if I could read the index *.ldb files it would be much easyer that linearize-data.py shots in the dark. Where can I find info on the index files?

**Comment by educob** [comment]
I have seen that just ignoring the 8 bytes of zeros everything works ok.

So the question is, why does the node, apparently randomly, write 8 bytes with zeros between blocks in blkxxxxx.dat?

**Comment by MarcoFalke** [comment]
Is this helper script still useful? If so, a test should probably be added to test it. Otherwise, it might be best to remove it.

**Comment by educob** [comment]
I changed:

```
            if (not inhdr or (inhdr[0] == "\0")):
                self.inF.close()
                self.inF = None
                self.inFn = self.inFn + 1
                continue
```

to:

```
            if (not inhdr or (inhdr[0] == "\0")):
                if(inhdr):
                    print("inhdr was 0. inFn:" + str(self.inFn) + str(len(inhdr)))
                    continue
                self.inF.close()
                self.inF = None
                self.inFn = self.inFn + 1
                continue
```

So it was useful. 
But I would like to know why the node insert these 8 zero bytes.
Thanks.

**Comment by mamcmanus** [comment]
@MarcoFalke yes, very useful. My node regularly has data corruption issues (still can't figure out why). Using -reindex or -reindex-chainstate does not always fix the issue, instead it usually just causes more data corruption issues. I run your script to make bootstrap.dat backup files which I use to reseed my node after it runs into issues. 

I don't know how to write tests yet but I will try to figure it out and add some tests for this script, which I think is invaluable. 

> Is this helper script still useful? If so, a test should probably be added to test it. Otherwise, it might be best to remove it.



@educob I have a similar issue but it occurs at blk00201.dat (about ~57% of the way through at height 575910 on today's date). I get this error:

`Invalid magic: 00000000`

I modified the inMagic if statement to continue on if it fails the check but equals '00000000', which has worked so far. Obviously not thrilled about this solution because I don't know why I had to implement it to begin with. Doesn't answer your question about the node randomly writing zeros but know that you are not alone!

> I have seen that just ignoring the 8 bytes of zeros everything works ok.
> 
> So the question is, why does the node, apparently randomly, write 8 bytes with zeros between blocks in blkxxxxx.dat?

**Comment by takinbo** [comment]
A few weeks ago, I was also having this same problem while attempting to export the block data. This problem was also discussed a few years ago in #5028. I worked on a slightly more elegant solution to search for the next position of the valid magic bytes rather than failing in #16802. I have ran the script with block data from a node with the issue and another without the issue and they both yield the same data output.
